src/ctypes/ldouble_stubs.c

```diff
@@ -0,0 +1,525 @@
+    CAMLparam2(a, b);                                                                 \
+    CAMLreturn(ctypes_copy_ldouble( ldouble_custom_val(a) OP ldouble_custom_val(b))); \
+  }
+
+OP2(add, +)
+OP2(sub, -)
+OP2(mul, *)
+OP2(div, /)
+
+CAMLprim value ctypes_ldouble_neg(value a) { 
+  CAMLparam1(a);
+  CAMLreturn(ctypes_copy_ldouble( - ldouble_custom_val(a))); 
+}
+
+#define FN1(OP)                                                   \
+  CAMLprim value ctypes_ldouble_ ## OP (value a) {                \
+    CAMLparam1(a);                                                \
+    CAMLreturn(ctypes_copy_ldouble( OP (ldouble_custom_val(a)))); \
+  }
+
+#define FN2(OP)                                                                          \
+  CAMLprim value ctypes_ldouble_ ## OP (value a, value b) {                              \
+    CAMLparam2(a, b);                                                                    \
+    CAMLreturn(ctypes_copy_ldouble( OP (ldouble_custom_val(a), ldouble_custom_val(b)))); \
+  }
+
+FN2(powl)
+FN1(sqrtl)
+FN1(expl)
+FN1(logl)
+FN1(log10l)
+FN1(expm1l)
+FN1(log1pl)
+FN1(cosl)
+FN1(sinl)
+FN1(tanl)
+FN1(acosl)
+FN1(asinl)
+FN1(atanl)
+FN2(atan2l)
+FN2(hypotl)
+FN1(coshl)
+FN1(sinhl)
+FN1(tanhl)
+FN1(acoshl)
+FN1(asinhl)
+FN1(atanhl)
+FN1(ceill)
+FN1(floorl)
+FN1(fabsl)
+FN2(remainderl)
+FN2(copysignl)
+
+#undef OP2
+#undef FN1
+#undef FN2
+
+CAMLprim value ctypes_ldouble_frexp(value v) {
+  CAMLparam1(v);
+  CAMLlocal2(r, rfv);
+  long double f = ldouble_custom_val(v);
+  int ri;
+  long double rf;
+  r = caml_alloc_tuple(2);
+  rf = frexpl(f, &ri);
+  rfv = ctypes_copy_ldouble(rf);
+  Store_field(r,0, rfv);
+  Store_field(r,1, Val_int(ri));
+  CAMLreturn(r);
+}
+
+CAMLprim value ctypes_ldouble_ldexp(value vf, value vi) {
+  CAMLparam2(vf, vi);
+  CAMLlocal1(r);
+  long double f = ldouble_custom_val(vf);
+  int i = Int_val(vi);
+  long double rf = ldexpl(f, i);
+  r = ctypes_copy_ldouble(rf);
+  CAMLreturn(r);
+}
+
+CAMLprim value ctypes_ldouble_modf(value v) {
+  CAMLparam1(v);
+  CAMLlocal1(r);
+  long double f = ldouble_custom_val(v);
+  long double rf2;
+  long double rf1 = modfl(f, &rf2);
+  r = caml_alloc_tuple(2);
+  Store_field(r, 0, ctypes_copy_ldouble(rf1));
+  Store_field(r, 1, ctypes_copy_ldouble(rf2));
+  CAMLreturn(r);
+}
+
+enum {
+  ml_FP_NORMAL = 0,
+  ml_FP_SUBNORMAL,
+  ml_FP_ZERO,
+  ml_FP_INFINITE,
+  ml_FP_NAN,
+};
+
+CAMLprim value ctypes_ldouble_classify(value v){
+  CAMLparam1(v);
+  CAMLlocal1(r);
+  long double f = ldouble_custom_val(v);
+  switch (fpclassify(f)){
+  case FP_NORMAL    : r = Val_int(ml_FP_NORMAL); break;
+  case FP_SUBNORMAL : r = Val_int(ml_FP_SUBNORMAL); break;
+  case FP_ZERO      : r = Val_int(ml_FP_ZERO); break;
+  case FP_INFINITE  : r = Val_int(ml_FP_INFINITE); break;
+  case FP_NAN       : 
+  default           : r = Val_int(ml_FP_NAN); break;
+  }
+  CAMLreturn(r);
+}
+
+static char *format_ldouble(char *fmt, long double d) {
```

Comment:
I'd be happier with an implementation that didn't use static storage here.  
There's a useful trick with `snprintf` to calculate the size: you can pass `NULL` and `0` as the first two arguments, and `snprintf `will give you back the size (sans terminating nul) needed to store the result.

So it should be sufficient to have a single call to `malloc`, two calls to `snprintf`, and no recursion.

(I think `snprintf` is broken on some older MSVC versions; in such cases I think it'd be fine to truncate rather than use an elaborate workaround.)